app/api/db/models/user_models.py

```diff
@@ -3,45 +3,44 @@
 from typing import Optional
 from uuid import UUID
 
-from sqlalchemy import Boolean, Column, Date, ForeignKey, Text
-from sqlalchemy.dialects.postgresql import UUID as postgresUUID
+from sqlalchemy import Boolean, Column, Date, Enum, ForeignKey, Text
+from sqlalchemy.dialects import postgresql
 from sqlalchemy.orm import Mapped, relationship
 
 import api.logging
 
-from .base import Base, TimestampMixin, uuid_gen
+from . import base
 
 logger = api.logging.get_logger(__name__)
 
 
-class RoleEnum(str, enum.Enum):
-    USER = "User"
-    ADMIN = "Admin"
-    THIRD_PARTY = "Third Party"
+class RoleType(str, enum.Enum):
+    USER = "USER"
+    ADMIN = "ADMIN"
 
 
-class User(Base, TimestampMixin):
+class User(base.Base, base.IdMixin, base.TimestampMixin):
     __tablename__ = "user"
 
-    user_id: Mapped[UUID] = Column(postgresUUID(as_uuid=True), primary_key=True, default=uuid_gen)
-
     first_name: str = Column(Text, nullable=False)
     middle_name: Optional[str] = Column(Text)
     last_name: str = Column(Text, nullable=False)
     phone_number: str = Column(Text, nullable=False)
     date_of_birth: date = Column(Date, nullable=False)
     is_active: bool = Column(Boolean, nullable=False)
 
-    roles: Optional[list["UserRole"]] = relationship(
-        "UserRole", uselist=True, back_populates="user"
-    )
+    roles: Optional[list["Role"]] = relationship("Role", back_populates="user")
 
 
-class UserRole(Base, TimestampMixin):
-    __tablename__ = "user_role"
+class Role(base.Base, base.TimestampMixin):
+    __tablename__ = "role"
     user_id: Mapped[UUID] = Column(
-        postgresUUID(as_uuid=True), ForeignKey("user.user_id"), primary_key=True
+        postgresql.UUID(as_uuid=True), ForeignKey("user.id"), primary_key=True
     )
-    role_description: str = Column(Text, primary_key=True, nullable=False, index=True)
-
+    # Set native_enum=False to use store enum values as VARCHAR/TEXT.
+    # This is more convenient and avoids needing create a new migration
+    # when adding a new enum value.
+    # https://docs.sqlalchemy.org/en/14/core/type_basics.html#sqlalchemy.types.Enum.params.native_enum
+    # https://medium.com/swlh/postgresql-3-ways-to-replace-enum-305861e089bc
```

Comment:
This article seems to suggest not doing it this way unless I'm misinterpreting it. Seems like it recommends native enum or foreign keys depending on the number of values

Reply:
I'm personally fine with native, I'll remove the native=False

Reply:
How does native work with regards to migrations? I think I remember not using those because alembic can't detect changes to enums so any changes require you to manually write it into alembic.

For example, use an enum and alembic should make it then alter it and run the migration create again and see what happens.

Reply:
@chouinar just tried removing native=False, and I think it's less optimal from a maintenance perspective, since alembic no longer seems to generate correct migrations.

without native=false, the first migration generated is
```
sa.Column("type", sa.Enum("USER", "ADMIN", name="roletype"), nullable=False),
```
then if i run the migration and then add another value to the enum e.g.
```
class RoleType(str, enum.Enum):
    USER = "USER"
    ADMIN = "ADMIN"
    THIRD_PARTY = "THIRD_PARTY"
```
then run migrate create, it creates an empty migration that just says
```
pass
```
for both upgrade and downgrade.

on the other hand with native=False, the initial migration is
```
sa.Column(
    "type", sa.Enum("USER", "ADMIN", name="roletype", native_enum=False), nullable=False
),
```
and then the subsequent migration after adding an enum value is
```
op.alter_column(
    "role",
    "type",
    existing_type=sa.VARCHAR(length=5),
    type_=sa.Enum("USER", "ADMIN", "THIRD_PARTY", name="roletype", native_enum=False),
    existing_nullable=False,
)
```
which looks correct

Reply:
@chouinar oh i totally missed your previous comment lol. I guess I independently confirmed what you already told me

Reply:
If I'm reading the non-native config right, it sounds like it still uses a check-constraint on the DB for the enum, just as a string? If so, then the non-native approach seems fine. Can't accidentally put bad data in. The "biggest" issue I see is that when doing DB queries there isn't a great way to know for certain all the possible values, but that's not the end of the world if you have access to the repo.

Reply:
Actually by default there's no check constraint
<img width="895" alt="image" src="https://user-images.githubusercontent.com/447859/208162031-2029053e-3fd7-4bb5-8153-ec1b14fcfd12.png">

we'd have to add `check_constraint=True` in which case i think you can see all the possible values by just doing `\d table` lemme try that out and paste here the results

Reply:
Tried it, and here's what I found.
Initial generated migration:
<img width="958" alt="image" src="https://user-images.githubusercontent.com/447859/208163644-bae6c341-36e2-411a-9190-71da9138588a.png">

And you can see all the possible enum values from looking in the db without looking at code by inspecting the table constraint
<img width="877" alt="image" src="https://user-images.githubusercontent.com/447859/208163729-43561e8d-052b-4594-80f8-7f71c77deef0.png">

And alembic successfully detects the change in enum
<img width="423" alt="image" src="https://user-images.githubusercontent.com/447859/208163803-5b3be717-1de5-4829-8ea3-81e2c7a48c9a.png">

Sounds like a good compromise to me, thoughts?

Reply:
Actually I spoke too soon, actually running the migration to modify the enum doesn't work out of the box
```
18:29:05.580  alembic.runtime.migration            run_migrations               INFO     Running upgrade 84f3065cba80 -> 3813d6d7f022, add third party role type          
Traceback (most recent call last):
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.DuplicateObject: constraint "role_`roletype_check`" for relation "role" already exists


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/srv/api/db/migrations/run.py", line 24, in up
    command.upgrade(alembic_cfg, revision)
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/command.py", line 322, in upgrade
    script.run_env()
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/script/base.py", line 569, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 94, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 110, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/srv/api/db/migrations/env.py", line 110, in <module>
    run_migrations_online()
  File "/srv/api/db/migrations/env.py", line 104, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/runtime/environment.py", line 853, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/runtime/migration.py", line 623, in run_migrations
    step.migration_fn(**kw)
  File "/srv/api/db/migrations/versions/2022_12_16_add_third_party_role_type.py", line 20, in upgrade
    op.alter_column(
  File "<string>", line 8, in alter_column
  File "<string>", line 3, in alter_column
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/operations/ops.py", line 1880, in alter_column
    return operations.invoke(alt)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/operations/base.py", line 399, in invoke
    return fn(self, operation)
           ^^^^^^^^^^^^^^^^^^^
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/operations/toimpl.py", line 74, in alter_column
    operations.impl.add_constraint(constraint)
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/ddl/impl.py", line 335, in add_constraint
    self._exec(schema.AddConstraint(const))
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/alembic/ddl/impl.py", line 195, in _exec
    return conn.execute(construct, multiparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1380, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/ddl.py", line 80, in _execute_on_connection
    return connection._execute_ddl(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1472, in _execute_ddl
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/root/.cache/pypoetry/virtualenvs/template-application-flask-o9msT97p-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.DuplicateObject) constraint "role_`roletype_check`" for relation "role" already exists

[SQL: ALTER TABLE role ADD CONSTRAINT "role_`roletype_check`" CHECK (type IN ('USER', 'ADMIN', 'THIRDPARTY'))]
(Background on this error at: https://sqlalche.me/e/14/f405)
make: *** [db-upgrade] Error 1
```

looks like alembic isn't smart enough to drop the constraint first before creating it new

Reply:
we could probably manually update the generated sql file to drop the constraint then add it back in, but i'm not sure that's worth it. i think SQLAlchemy already does validation at the code level, i feel like it might be more work than it's worth to add the constraint at the db level. projects can probably do that on their own if they feel the need to.

Reply:
Pushed an update to the comment with a better explanation of why we chose native=false and constraint=false

Reply:
Yeah, I think as-is is probably best. Enums with alembic boil down to code-only without it being a pain to manage. I think alembic lets you make your own utilities to build into it that technically it's possible to make these a proper check-constraint, but just not worth it. The only benefit is when you do need to work on raw-sql which from the projects I've been on tend to be just select queries and not inserts (outside of fixing data).